core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -329,42 +331,40 @@ private void applyConfiguration(CreateContainerCmd createCommand) {
     }
 
     /**
-     * Wait until the container has started. The default implementation simply
-     * waits for a port to start listening; subclasses may override if more
-     * sophisticated behaviour is required.
+     * Specify the {@link WaitStrategy} to use to determine if the container is ready.
+     *
+     * @see Wait#defaultWaitStrategy()
+     * @param waitStrategy the WaitStrategy to use
+     * @return the {@code waitStrategy}, for chaining
      */
-    protected void waitUntilContainerStarted() {
-        waitForListeningPort(DockerClientFactory.instance().dockerHostIpAddress(), getLivenessCheckPort());
+    public WaitStrategy waitingFor(WaitStrategy waitStrategy) {
```

Comment:
Could you make this return GenericContainer instead of the wait strategy? This way it'd be possible to do:

```
@ClassRule
public static GenericContainer elasticsearch =
    new GenericContainer("elasticsearch:2.3")
               .withExposedPorts(9200)
               .waitingFor(new HttpWaitStrategyImpl()
                   .waitForStatusCode("/_all", 301)
                   .usingTls())
               .withSomeOtherGenericContainerBuilderMethod(...) ... 
```